src/Microsoft.OData.Core/Uri/NodeToStringBuilder.cs

```diff
@@ -470,7 +471,197 @@ internal String TranslateNode(QueryNode node)
         internal String TranslateFilterClause(FilterClause filterClause)
         {
             Debug.Assert(filterClause != null, "filterClause != null");
-            return this.TranslateNode(filterClause.Expression);
+            string translatedNode = this.TranslateNode(filterClause.Expression);
+
+            switch (filterClause.Expression.Kind)

```

Comment:
can we refactor the implementation here?
1) separate the whole into methods
2) Use the traverse, for example: https://github.com/OData/WebApi/blob/master/src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs#L184

Reply:
@xuzhg Done